p2p/net/swarm/swarm.go

```diff
@@ -390,6 +397,19 @@ func (s *Swarm) addConn(tc transport.CapableConn, dir network.Direction) (*Conn,
 	c.notifyLk.Lock()
 	s.conns.Unlock()
 
+	// Notify goroutines waiting for a direct connection
+
+	// Go routines interested in waiting for direct connection first acquire this lock and then
+	// acquire conns.RLock. Do not acquire this lock before conns.Unlock to prevent deadlock.
+	s.directConnNotifs.Lock()
+	if !c.Stat().Transient {
```

Comment:
Any reason not to move the locking of the mutex into the if? That way we avoid locking altogether for non-transient conns.

Reply:
Fixed this. Moved the mutex lock in to the if statement. We will not lock for transient connections. We need to lock for direct connections since we need to notify waiters.